### ISSUE TITLE
netty dns resolving performance sometimes decreased greatly

### ISSUE THREAD

**Issue by jianyuanzh** [netty/netty#8013]
### Expected behavior
We have used netty dns resolver as our product's dns resolver. And from the metrics monitored by our monitoring tool, we have about 130+ resolving per seconds:
![image](https://user-images.githubusercontent.com/12515529/41154799-e653c120-6b4d-11e8-8851-92d0626757b7.png)
And normally the resolving performance is great, and it is about 50ms per resolving.
![image](https://user-images.githubusercontent.com/12515529/41154838-1463f4f4-6b4e-11e8-9724-5234a0b07547.png)

### Actual behavior
But the situation may get worse and become unrecoverable unless we restart our java process. Average resolving time may be increased to more than 4 seconds...  

It will be very helpful if you guys have any ideas for this case.  

Thanks

### Steps to reproduce
It happens occasionally

### Netty version
io.netty:netty-all:4.1.18.Final

### JVM version (e.g. `java -version`)
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

### OS version (e.g. `uname -a`)

```
Linux production 2.6.32-696.3.2.el6.x86_64 #1 SMP Tue Jun 20 01:26:55 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by normanmaurer** [comment]
Can you provide some more details... Like can you share a heap dump and thread-dump on when this happens before you bounce the service ?

**Comment by jianyuanzh** [comment]
@normanmaurer  will do this next time this happens again.

**Comment by johnou** [comment]
Are you also monitoring the JVM heap? Sounds like you might be seeing delays during full GC collection.

**Comment by jianyuanzh** [comment]
@johnou , yes, we also monitors JVM heap. All memory usage, GC metrics were all OK around that time:
![image](https://user-images.githubusercontent.com/12515529/41461333-0277533e-70c2-11e8-807e-aeee2e982c49.png)
![image](https://user-images.githubusercontent.com/12515529/41461362-16b0f9d6-70c2-11e8-8142-23a467c6aaf1.png)
![image](https://user-images.githubusercontent.com/12515529/41461401-2e78ffc8-70c2-11e8-90ce-605bae3e6fa3.png)

So, only one full GC and used time was small. 

Thanks

**Comment by jianyuanzh** [comment]
@normanmaurer , We observed this issue again today. And this time we get more details.

Some DNS resolve task failed with the exception "query ID space exhausted". From netty code, we know this happens when there is too much DNS concurrency.  The workaround for fixing this is either "renew a netty dns resolver" or clear the related cache in memory. 

Here is the statistic for `DnsQueryContextManager#map`:
```
dnsServer: cacheCount
/8.8.4.4:53  :  0
/0:0:0:0:0:0:808:404:53  :  0
/216.52.126.1:53  :  13681
/0:0:0:0:0:0:d834:7e01:53  :  13681
/216.52.254.1:53  :  615
/216.52.254.33:53  :  1
/0:0:0:0:0:0:d834:fe01:53  :  615
/8.8.8.8:53  :  0
/0:0:0:0:0:0:808:808:53  :  0
/0:0:0:0:0:0:d834:fe21:53  :  1
/127.0.0.1:53  :  65536
localhost/0:0:0:0:0:0:0:1:53  :  65536
```
The cached DnsQueryContext count reaches the up limit for the localhost dns server. We cannot have that huge resolving tasks in concurrent.  We suspect the `DnsQueryContext` item is not properly removed in some edge cases. 

Seems the `remove` method is invoked by `DnsQueryContext#setSucces` and `DnsQueryContext#setFailure`.  

Can you help to find is there any branches which also need remove the `DnsQueryContext` from the cache ?

Thanks!

**Comment by jianyuanzh** [comment]
I did not dump the heap this time. But I can still see the cache size keep increasing slowly.  I will try to take a dump tomorrow.

**Comment by jianyuanzh** [comment]
@normanmaurer , The root cause should be the DNS server did not respond to the query at all.
 
I compared the increasing items with the network packages and verified that the un-released items had no related response. 

See given example, the 2987 item is added in the cache around 09:13 and cannot be released. From the related 3 minutes tcpdump data, I can see the 2987 query request but no related response:

  
![image](https://user-images.githubusercontent.com/12515529/41863588-d43c16f6-78d9-11e8-8756-aac6a6822661.png)


I think netty resolver can improve the DNS query context cache to not hold the query context in memory for too long time, eg: introduce some expiry

**Comment by normanmaurer** [comment]
@jianyuanzh hmm... the query should have been timeout after 5 seconds by default. Can you show me how you configure your `DnsNameResolver` ?

**Comment by jianyuanzh** [comment]
@normanmaurer , yes, there is the timeout , but the timeout handler will not try remove the context if the promise is canceled. And the promise in memory is under canceled status.

```
 private void onQueryWriteCompletion(ChannelFuture writeFuture) {
        if (!writeFuture.isSuccess()) {
            setFailure("failed to send a query", writeFuture.cause());
            return;
        }

        // Schedule a query timeout task if necessary.
        final long queryTimeoutMillis = parent.queryTimeoutMillis();
        if (queryTimeoutMillis > 0) {
            timeoutFuture = parent.ch.eventLoop().schedule(new Runnable() {
                @Override
                public void run() {
                    if (promise.isDone()) { // the one in memory isDone and isCanceled.
                        // Received a response before the query times out.
                        return;
                    }

                    setFailure("query timed out after " + queryTimeoutMillis + " milliseconds", null);
                }
            }, queryTimeoutMillis, TimeUnit.MILLISECONDS);
        }
    }
```

**Comment by normanmaurer** [comment]
@jianyuanzh ah this makes sense then... Can you check https://github.com/netty/netty/pull/8058 ?

**Comment by normanmaurer** [comment]
Fix was merged into 4.1
